### PR TITLE
fix(core): fix DB crashes while writing to multiple tables 

### LINKED ISSUE #3062: Occasional SIGSEGV on 7.0.1
### Describe the bug

I'm seeing the database crash occasionally when doing lots of inserts using the Java ILP client

```
tsMax=2023-03-10T18:06:17.807000Z, commitToTimestamp=1969-12-31T23:59:59.999999Z]
2023-03-10T18:06:18.506628Z I i.q.c.TableWriter processing WAL [path=/var/lib/questdb/db/controller-status~8/wal7/0, roLo=35642, seqTxn5951, roHi=35675, tsMin=2023-03-10T18:06:16.798000Z, tsMax=2023-03-10T18:06:17.818000Z, commitToTimestamp=1969-12-31T23:59:59.999999Z]
2023-03-10T18:06:18.514509Z I i.q.c.TableWriter processing WAL [path=/var/lib/questdb/db/controller-status~8/wal6/0, roLo=35991, seqTxn5952, roHi=36023, tsMin=2023-03-10T18:06:16.801000Z, tsMax=2023-03-10T18:06:17.806000Z, commitToTimestamp=294247-01-10T04:00:54.775807Z]
2023-03-10T18:06:18.515401Z I i.q.c.TableWriter sorting WAL [table=controller-status, ordered=false, lagRowCount=258, rowLo=35991, rowHi=36023]
2023-03-10T18:06:18.583126Z I i.q.c.TableWriter closing last partition [table=device-status]
2023-03-10T18:06:18.587159Z I i.q.c.TableWriter merged partition [table=`device-status`, ts=2023-03-10T18:00:00.000000Z, txn=3674]
2023-03-10T18:06:18.592116Z I i.q.c.TableWriter switched partition [path='/var/lib/questdb/db/device-status~5/2023-03-10T18.3674']
2023-03-10T18:06:18.669128Z I i.q.c.TableWriter o3 partition task [table=controller-status, partitionIsReadOnly=false, srcOooBatchRowSize=290, srcOooLo=0, srcOooHi=289, srcOooMax=290, o3RowCount=290, o3LagRowCount=0, srcDataMax=45829, o3TimestampMin=2023-03-10T18:06:15.602000Z, o3Timestamp=2023-03-10T18:06:15.602000Z, o3TimestampMax=2023-03-10T18:06:17.818000Z, partitionTimestamp=2023-03-10T18:00:00.000000Z, partitionIndex=4, partitionSize=46119, maxTimestamp=2023-03-10T18:06:15.643000Z, last=true, append=false, pCount=1, flattenTimestamp=true, memUsed=108346810934]
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f2a2c0baa08, pid=1, tid=91
#
# JRE version: OpenJDK Runtime Environment Corretto-17.0.3.6.1 (17.0.3+6) (build 17.0.3+6-LTS)
# Java VM: OpenJDK 64-Bit Server VM Corretto-17.0.3.6.1 (17.0.3+6-LTS, mixed mode, tiered, compressed oops, compressed class ptrs, g1 gc, linux-amd64)
# Problematic frame:
# C  [libquestdb16208632089475074569.so+0x6fa08]  merge_copy_var_column_int32_AVX2(index_t*, long, long*, char*, long*, char*, long*, char*, long)+0xa8
#
# Core dump will be written. Default location: /var/lib/questdb/core.1
#
# An error report file with more information is saved as:
# /var/lib/questdb/db/hs_err_pid+1.log
#
# If you would like to submit a bug report, please visit:
#   https://github.com/corretto/corretto-17/issues/
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
```

### To reproduce

_No response_

### Expected Behavior

_No response_

### Environment

```markdown
- **QuestDB version**: 7.0.1 (docker image)
- **OS**: Amazon Linux, m6a.16xlarge instance
- **Browser**: N/A (Java ILP client inserts)
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ideoma**
Should fix #3062

Other fixes:
- commit WAL seqTxn with column add / remove transaction to prevent WAL table be suspended after standalone ALTER TABLE transaction.
- fix lag size calculation for WAL apply

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 310 / 351 (88.32%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|266|306|86.93%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/vect/GroupByNotKeyedVectorRecordCursorFactory.java|12|13|92.31%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/tasks/O3CallbackTask.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3CallbackJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/vect/GroupByJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/tasks/O3CopyTask.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/vect/VectorAggregateEntry.java|15|15|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/ApplyWal2TableJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/vect/GroupByRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/AbstractMultiTenantPool.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/mp/SOUnboundedCountDownLatch.java|1|1|100.00%|

File: core/src/main/java/io/questdb/cairo/TableWriter.java [hunk 1]
```diff
@@ -3828,7 +3906,6 @@ private void o3MergeIntoLag(long mergedTimestamps, long countInLag, long mappedR
                                 mappedRoHi,
                                 ColumnType.isVariableLength(type) ? o3MergeLagVarColumnRef : o3MergeLagFixColumnRef
                         );
-                        o3PendingCallbackTasks.add(task);
                     } finally {
```
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/cairo/TableWriter.java:3909]: Do we really need this `try/finally` block here? I don't think anything in the `try` block may throw. The same question applies to other similar places where we initialize `O3CallbackTask`s.

File: core/src/main/java/io/questdb/cairo/TableWriter.java [hunk 2]
```diff
@@ -3007,6 +3026,22 @@ private boolean createWalSymbolMapping(SymbolMapDiff symbolMapDiff, int columnIn
...
+        final Sequence subSeq = this.messageBus.getO3CallbackSubSeq();
+        while (!o3DoneLatch.done(queuedCount)) {
+            long cursor = subSeq.next();
+            if (cursor > -1) {
+                O3CallbackJob.runCallbackWithCol(queue.get(cursor), cursor, subSeq);
+            }
+
+            if (cursor == -1) {
+                o3DoneLatch.await(queuedCount);
+            }
```
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/cairo/TableWriter.java:3040]: nit: would be nice to add a backoff here similar to what we do in other such places:
```java
if (cursor == -2) {
    Os.pause();
}
```

File: core/src/main/java/io/questdb/cairo/TableWriter.java [hunk 3]
```diff
@@ -4273,18 +4386,10 @@ private long o3ScheduleMoveUncommitted0(int timestampIndex, long transientRowsAd
...
-                            -1,
-                            null
-                    );
-                }
-            }
-
-            o3DoneLatch.await(queuedCount);
+            dispatchO3CallbackQueue(queue, queuedCount);
+        }
+        if (o3ErrorCount.get() > 0) {
```
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/cairo/TableWriter.java:4391]: Do we need this check here considering that `dispatchO3CallbackQueue` has a `checkO3Errors` call?

File: core/src/main/java/io/questdb/griffin/engine/groupby/vect/GroupByNotKeyedVectorRecordCursorFactory.java [hunk 4]
```diff
@@ -109,26 +107,37 @@ public boolean usesCompiledFilter() {
...
+                VectorAggregateTask task = queue.get(cursor);
+                task.entry.run(workerId, subSeq, cursor);
                 reclaimed++;
             }
-        }
 
-        log.info().$("waiting for parts [queuedCount=").$(queuedCount).$(']').$();
-        doneLatch.await(queuedCount);
+            if (cursor == -1) {
+                if (circuitBreaker.checkIfTripped()) {
```
[the remark this answers is not in the document]
Comment (ideoma) [reply core/src/main/java/io/questdb/griffin/engine/groupby/vect/GroupByNotKeyedVectorRecordCursorFactory.java]: removed

File: core/src/main/java/io/questdb/griffin/engine/groupby/vect/GroupByNotKeyedVectorRecordCursorFactory.java [hunk 5]
```diff
@@ -109,26 +107,37 @@ public boolean usesCompiledFilter() {
...
+                    sharedCB.cancel();
+                }
+
+                log.info().$("waiting for parts [queuedCount=").$(queuedCount).$(']').$();
+                doneLatch.await(queuedCount);
+            }
```
[the remark this answers is not in the document]
Comment (ideoma) [reply core/src/main/java/io/questdb/griffin/engine/groupby/vect/GroupByNotKeyedVectorRecordCursorFactory.java:132]: added

File: core/src/main/java/io/questdb/cairo/TableWriter.java (hunk 1 above)
Comment (ideoma) [reply core/src/main/java/io/questdb/cairo/TableWriter.java]: removed

File: core/src/main/java/io/questdb/cairo/TableWriter.java (hunk 2 above)
Comment (ideoma) [reply core/src/main/java/io/questdb/cairo/TableWriter.java:3040]: added

File: core/src/main/java/io/questdb/cairo/TableWriter.java (hunk 3 above)
Comment (ideoma) [reply core/src/main/java/io/questdb/cairo/TableWriter.java]: removed

**Review by puzpuzpuz — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 317 / 354 (89.55%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|275|311|88.42%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/vect/GroupByNotKeyedVectorRecordCursorFactory.java|10|11|90.91%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/tasks/O3CallbackTask.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3CallbackJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/vect/GroupByJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/tasks/O3CopyTask.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/vect/VectorAggregateEntry.java|15|15|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/ApplyWal2TableJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/vect/GroupByRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/AbstractMultiTenantPool.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/mp/SOUnboundedCountDownLatch.java|1|1|100.00%|

**Comment by ideoma** [comment]
There is a nontrivial lag calculation bug which is completely rewritten in #3053 
I am closing this PR, it the changes will be merged to #3053

**Comment by ideoma** [comment]
Having thought of that, now that I know that the last fuzz failure test will be fixed in #3053, it still make sense to have this fix first.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 317 / 353 (89.80%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|274|310|88.39%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/vect/GroupByNotKeyedVectorRecordCursorFactory.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/tasks/O3CallbackTask.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/O3CallbackJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/vect/GroupByJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/tasks/O3CopyTask.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/vect/VectorAggregateEntry.java|15|15|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/ApplyWal2TableJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/vect/GroupByRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/AbstractMultiTenantPool.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/mp/SOUnboundedCountDownLatch.java|1|1|100.00%|

**Review by puzpuzpuz — APPROVED** [review]

**Pull request merged**